src/Runner.Worker/Container/ContainerHooks/ContainerHookManager.cs

```diff
@@ -0,0 +1,161 @@
+using System.Linq;
+using System.Threading.Tasks;
+using GitHub.DistributedTask.Pipelines.ContextData;
+using GitHub.DistributedTask.WebApi;
+using GitHub.Runner.Common;
+using GitHub.Runner.Common.Util;
+using GitHub.Runner.Sdk;
+using GitHub.Runner.Worker.Handlers;
+using GitHub.Services.WebApi;
+
+namespace GitHub.Runner.Worker.Container.ContainerHooks
+{
+    [ServiceLocator(Default = typeof(ContainerHookManager))]
+    public interface IContainerHookManager : IRunnerService
+    {
+        Task PrepareJobAsync(IExecutionContext context, List<ContainerInfo> containers);
+        Task CleanupJobAsync(IExecutionContext context, List<ContainerInfo> containers);
+        Task ContainerStepAsync(IExecutionContext context);
+        Task RunScriptStepAsync(IExecutionContext context);
+    }
+
+    public class ContainerHookManager : RunnerService, IContainerHookManager
+    {
+        private const string ResponseFolderName = "_hook_responses";
+        private string HookIndexPath;
+        public override void Initialize(IHostContext hostContext)
+        {
+            base.Initialize(hostContext);
+            HookIndexPath = $"{Environment.GetEnvironmentVariable(Constants.Hooks.ContainerHooksPath)}";
+        }
+        
+        public async Task PrepareJobAsync(IExecutionContext context, List<ContainerInfo> containers)
+        {
+            Trace.Entering();
+            var responsePath = GenerateResponsePath();
+            var jobContainer = containers.Where(c => c.IsJobContainer).FirstOrDefault();
+            var serviceContainers = containers.Where(c => c.IsJobContainer == false).ToList();
+
+            var input = new HookInput
+            {
+                Command = HookCommand.PrepareJob,
+                ResponseFile = responsePath,
+                Args = new HookArgs
+                {
+                    JobContainer = jobContainer.GetHookContainer(),
+                    Services = serviceContainers.Select(c => c.GetHookContainer()).ToList(),
+                }
+            };            
+
+            var response = await ExecuteHookScript(context, input);
+            context.JobContext.Container["id"] = new StringContextData(response.Context.Container.Id);
+            jobContainer.ContainerId = response.Context.Container.Id;
+            context.JobContext.Container["network"] = new StringContextData(response.Context.Container.Network);
+            jobContainer.ContainerNetwork = response.Context.Container.Network;
+            context.JobContext["hook_state"] = new StringContextData(JsonUtility.ToString(response.State));
+            // var configEnvFormat = "--format \"{{range .Config.Env}}{{println .}}{{end}}\"";
+            // var containerEnv = await _dockerManager.DockerInspect(executionContext, container.ContainerId, configEnvFormat);
+            // container.ContainerRuntimePath = DockerUtil.ParsePathFromConfigEnv(containerEnv);
+
+            for (var i = 0; i < response.Context.Services.Count; i++)
```

Comment:
same concern about this throwing an exception if context isn't set

Reply:
Let's clarify above if we want to enforce context